nerdlets/maturity-nerdlet/tags.js

```diff
@@ -0,0 +1,154 @@
+import React from 'react';
+import { Dropdown, DropdownSection, DropdownItem, NerdGraphQuery } from 'nr1';
+
+export default class Tags extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      allTags: null,
+      searchText: '',
+      selectedTag: null,
+      tagTitle: 'Tags'
+    }
+  }
+
+  async componentDidMount() {
+    await this.constructTags();
+  }
+
+  async constructTags(){
+    let domains = ['APM', 'INFRA', 'SYNTH']; //start with these domains for sample set
+    let allTags = {};
+    let proms = [];
+
+    for (let domain of domains) {
+      let q = `
+      {
+        actor {
+          entitySearch(queryBuilder: {domain: ${domain}}) {
+            results {
+              entities {
+                tags {
+                  key
+                  values
+                }
+              }
+            }
+          }
+        }
+      }
+      `
+      proms.push(this.queryTags(q));
+    }
+
+    Promise.all(proms).then(resp => {
+      let allTags = [];
+      let flattend = resp.flat();
+      let first = true;
+
+      allTags.push({title: 'Default', items: ['Default']}); //Default option
+
+      let i = 1;
+      for (let entityTags of flattend) {
+        if (first == true) {
+          for (let aTag of entityTags.tags) {
+            allTags.push({title: aTag.key, items: []});
+            allTags[i].items.push(aTag.values[0]);
+            i++;
+          }
+          first = false;
+        } else {
+          for (let anotherTag of entityTags.tags) {
+            let obj = allTags.filter(o => {
+              return o.title == anotherTag.key;
+            })
+            if (obj.length > 0) {
+              if (obj[0].items.includes(anotherTag.values[0])) {
+                //do nothing
+              } else {
+                obj[0].items.push(anotherTag.values[0]);
+              }
+            } else {
+              allTags.push({title: anotherTag.key, items: []});
+              allTags[i].items.push(anotherTag.values[0]);
+            }
+          }
+        }
+      }
+
+      this.setState({allTags: allTags});
+    })
+  }
+
+  async queryTags(q) {
+    let res = await NerdGraphQuery.query({query: q});
+    if (res.errors) {
+      console.debug(res.errors);
```

Comment:
Error handling here needs to be modified. If NerdGraph returns an error, this function will cause a `TypeError: Cannot read property 'tags' of undefined` on line 54 since `entityTags` will be undefined.

Reply:
should be fixed in my latest commit